File: src/main/java/at/yawk/wm/dock/module/widget/TagListWidget.java
package at.yawk.wm.dock.module.widget;

import at.yawk.wm.style.FontStyle;
import at.yawk.wm.ui.Direction;
import at.yawk.wm.ui.FlowCompositeWidget;
import at.yawk.wm.ui.TextWidget;
import at.yawk.wm.dock.module.DockConfig;
import at.yawk.wm.dock.module.DockWidget;
import at.yawk.wm.dock.module.FontSource;
import at.yawk.wm.ui.RenderElf;
import at.yawk.wm.hl.HerbstClient;
import at.yawk.wm.hl.HerbstEventBus;
import at.yawk.wm.hl.Monitor;
import at.yawk.wm.hl.Tag;
import java.util.ArrayList;
import java.util.List;
import javax.inject.Inject;

/**
 * @author yawkat
 */
@DockWidget(position = DockWidget.Position.LEFT)
public class TagListWidget extends FlowCompositeWidget {
    private final List<TextWidget> tagWidgets = new ArrayList<>();

    @Inject FontSource fontSource;
    @Inject HerbstClient herbstClient;
    @Inject RenderElf renderElf;
    @Inject HerbstEventBus herbstEventBus;
    @Inject Monitor monitor;

    @Override
    public void init() {
        update();
        herbstEventBus.addTagEventHandler(event -> update());
    }

    private void update() {
        int i = 0;
        List<Tag> tags = herbstClient.getTags(monitor);
        for (; i < tags.size(); i++) {
            Tag tag = tags.get(i);
            TextWidget widget;
            if (tagWidgets.size() > i) {
                widget = tagWidgets.get(i);
            } else {
                widget = new TextWidget();
                if (tagWidgets.isEmpty()) {
                    widget.after(getAnchor(), Direction.HORIZONTAL);
                } else {
                    widget.after(tagWidgets.get(tagWidgets.size() - 1), Direction.HORIZONTAL);
                }
                addWidget(widget);
                tagWidgets.add(widget);
            }

            FontStyle style;
            switch (tag.getState()) {
            case SELECTED:
                style = DockConfig.INSTANCE.getActiveFont();
                break;
            case SELECTED_ELSEWHERE:
                style = DockConfig.INSTANCE.getActiveElsewhereFont();
                break;
            case RUNNING:
            default:
                style = DockConfig.INSTANCE.getRunningFont();
                break;
            case EMPTY:
                style = DockConfig.INSTANCE.getEmptyFont();
                break;
            }

            widget.setFont(fontSource.getFont(style));
            widget.setText(tag.getId());
        }
        renderElf.render();
    }
}

File: src/main/java/at/yawk/wm/dock/module/widget/TagListWidget.kt
package at.yawk.wm.dock.module.widget

import at.yawk.wm.dock.module.DockConfig
import at.yawk.wm.dock.module.DockWidget
import at.yawk.wm.dock.module.FontSource
import at.yawk.wm.hl.HerbstClient
import at.yawk.wm.hl.HerbstEventBus
import at.yawk.wm.hl.Monitor
import at.yawk.wm.hl.Tag
import at.yawk.wm.ui.Direction
import at.yawk.wm.ui.FlowCompositeWidget
import at.yawk.wm.ui.RenderElf
import at.yawk.wm.ui.TextWidget
import java.util.ArrayList
import javax.inject.Inject

@DockWidget(position = DockWidget.Position.LEFT)
class TagListWidget @Inject constructor(
    private val fontSource: FontSource,
    private val herbstClient: HerbstClient,
    private val renderElf: RenderElf,
    private val herbstEventBus: HerbstEventBus,
    private val monitor: Monitor,
) : FlowCompositeWidget() {

    private val tagWidgets: MutableList<TextWidget> = ArrayList()
    override fun init() {
        update()
        herbstEventBus.addTagEventHandler { update() }
    }

    private fun update() {
        var i = 0
        val tags = herbstClient.getTags(monitor)
        while (i < tags.size) {
            val tag = tags[i]
            var widget: TextWidget
            if (tagWidgets.size > i) {
                widget = tagWidgets[i]
            } else {
                widget = TextWidget()
                if (tagWidgets.isEmpty()) {
                    widget.after(anchor, Direction.HORIZONTAL)
                } else {
                    widget.after(tagWidgets[tagWidgets.size - 1], Direction.HORIZONTAL)
                }
                addWidget(widget)
                tagWidgets.add(widget)
            }
            val style = when (tag.state) {
                Tag.State.SELECTED -> DockConfig.activeFont
                Tag.State.SELECTED_ELSEWHERE -> DockConfig.activeElsewhereFont
                Tag.State.RUNNING -> DockConfig.runningFont
                Tag.State.EMPTY -> DockConfig.emptyFont
                else -> DockConfig.runningFont
            }
            widget.font = fontSource.getFont(style)
            widget.text = tag.id
            i++
        }
        renderElf.render()
    }
}